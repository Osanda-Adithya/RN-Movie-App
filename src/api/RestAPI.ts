import { ResponseTopRelated } from "../utils/ResponseModel"

const URL = "https://api.themoviedb.org/3/movie/top_rated?api_key=9618b5cf6ae9661f92fff553c697bed4&language=en-US&page="

export const RestAPI = (pageNo: string): Promise<ResponseTopRelated | undefined> => {
    return new Promise((resolve, reject) => {
        fetch(URL + pageNo, {
            method: 'GET'
        })
            .then(response => response.json())
            .then(res => {
                return resolve(res);
            }).catch(error => {
                return reject(error);
            })
    })

}