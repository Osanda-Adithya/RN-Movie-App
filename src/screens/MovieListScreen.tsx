import React, { FC, useEffect, useState } from "react";
import type { NativeStackScreenProps } from '@react-navigation/native-stack';
import { FlatList, View, Text, StyleSheet, Image, RefreshControl } from "react-native";
import { RestAPI } from "../api/RestAPI";
import { RootParamList } from "../navigator/StackNavigation";
import { Results } from "../utils/ResponseModel";
import { ternary } from "../utils/Tools";
import PrecentageChart from "../components/PrecentageChart";

type MovieListScreenProps = NativeStackScreenProps<RootParamList, 'MovieListScreen'>;

const MovieListScreen: FC<MovieListScreenProps> = () => {

    const [topRelated, setTopRelated] = useState<Results[]>([]);
    const [data, setData] = useState<{ isLoading: boolean, pageNo: number, rederID: number }>({ isLoading: true, pageNo: 1, rederID: 0 })

    useEffect(() => {
        fetchAPIData()
    }, [data.rederID])

    function fetchAPIData() {
        RestAPI(data.pageNo.toString()).then(res => {
            if (res) {
                setTopRelated(topRelated.concat(res.results));
                setData({ ...data, isLoading: false })
            }
        });
    }

    const renderSigleRow = ({ item, index }: { item: Results, index: number }) => {
        return (
            <View style={styles.singleRowContainer}>

                <Image source={{ uri: 'https://image.tmdb.org/t/p/w500' + item?.poster_path, width: 150, height: 200 }} style={styles.image} />

                <View style={styles.chartContainer}>
                    <PrecentageChart
                        color={"#04d11c"}
                        shadowColor={"#999"}
                        bgColor={"#042444"}
                        radius={25}
                        borderWidth={5}
                        percent={item.popularity}
                    />
                </View>

                <View style={styles.container}>
                    <Text numberOfLines={2} style={styles.originalTitle}>#{index + 1} {item?.original_title}</Text>
                    <Text style={styles.title}>({item?.title})</Text>
                    <View style={styles.subContainer}>
                        <Text style={styles.fontStyle}> {item?.release_date}  ({item?.original_language})</Text>
                        <View style={styles.dot} />
                        <View style={styles.genderFontContainer}>
                            <Text style={styles.fontStyle}>{ternary(item?.adult, "R", "All")}</Text>
                        </View>
                    </View>
                    <Text style={styles.fontStyle} numberOfLines={4}>{item?.overview}</Text>
                </View>

            </View>
        )
    }

    const onHandleEndReach = () => {
        setData({ isLoading: true, pageNo: data.pageNo + 1, rederID: new Date().getUTCMilliseconds() })
    }

    const onRefreshHandle = () => {
        setTopRelated([]);
        setData({ isLoading: true, pageNo: 1, rederID: new Date().getUTCMilliseconds() })
    }

    return (
        <View style={{ flex: 1 }}>
            <FlatList
                testID="flatlist_testid"
                data={topRelated}
                keyExtractor={(_, index) => index?.toString()}
                renderItem={renderSigleRow}
                onEndReached={onHandleEndReach}
                onEndReachedThreshold={0.5}
                refreshControl={
                    <RefreshControl
                        refreshing={data.isLoading}
                        colors={['red', 'green', 'blue']}
                        title={'Refreshing'}
                        titleColor={'black'}
                        onRefresh={onRefreshHandle}
                    />
                }
            />
        </View>
    )
}

export default MovieListScreen;

const styles = StyleSheet.create({
    singleRowContainer: {
        flex: 1,
        flexDirection: 'row',
        marginHorizontal: 10,
        marginVertical: 12,
        paddingVertical: 18,
        paddingHorizontal: 10,
        borderRadius: 10,
        backgroundColor: '#fff',
        elevation: 10
    },
    container: {
        flex: 1,
        marginVertical: 10,
        marginRight: 5,
        marginLeft: 14,
        justifyContent: 'center'
    },
    subContainer: {
        flexDirection: 'row',
        alignItems: 'center',
        marginVertical: 5
    },
    genderFontContainer: {
        width: 22,
        borderWidth: 1,
        borderColor: '#878686',
        borderRadius: 4,
        alignItems: 'center',
    },
    fontStyle: {
        fontSize: 14,
        color: '#878686'
    },
    dot: {
        width: 5,
        height: 5,
        borderRadius: 5,
        backgroundColor: '#878686',
        marginTop: 5,
        marginHorizontal: 10,
    },
    title: {
        fontSize: 16,
        fontWeight: 'bold',
        color: '#878686'
    },
    originalTitle: {
        fontSize: 18,
        fontWeight: 'bold',
        color: 'black'
    },
    image: {
        borderRadius: 10
    },
    chartContainer: {
        width: 40,
        height: 30,
        borderRadius: 50,
        position: 'absolute',
        top: 192,
        left: 64,
        alignItems: 'center',
        justifyContent: 'center',
    },
    precentage: {
        color: 'white',
        fontSize: 12
    }
})