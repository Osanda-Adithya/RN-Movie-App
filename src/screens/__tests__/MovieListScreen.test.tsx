import { act, fireEvent, render } from "@testing-library/react-native";
import React from "react";
import * as API from "../../api/RestAPI";
import MovieListScreen from "../MovieListScreen";

const component = (
    <MovieListScreen
        navigation={{ navigate: jest.fn() } as any}
        route={jest.fn() as any}
    />
)

const fakeResponse = {
    adult: false,
    backdrop_path: "/lXhgCODAbBXL5buk9yEmTpOoOgR.jpg",
    genre_ids: [
        12,
        14,
        28
    ],
    id: 122,
    original_language: "en",
    original_title: "The Lord of the Rings: The Return of the King",
    overview: "Aragorn is revealed as the heir to the ancient kings as he, Gandalf and the other members of the broken fellowship struggle to save Gondor from Sauron's forces. Meanwhile, Frodo and Sam take the ring closer to the heart of Mordor, the dark lord's realm.",
    popularity: 132.794,
    poster_path: "/rCzpDGLbOoPwLjy3OAm5NUPOTrC.jpg",
    release_date: "2003-12-01",
    title: "The Lord of the Rings: The Return of the King",
    video: false,
    vote_average: 8.5,
    vote_count: 20340
};

describe('render successfully', () => {
    it('render snapshot', () => {
        const container = render(component);
        expect(container.toJSON()).toMatchSnapshot();
    })
    it('flatlist check', () => {
        const container = render(component);
        fireEvent(container.getByTestId('flatlist_testid'), 'keyExtractor');
    })
    it('flatlist onEndReached check', () => {
        const container = render(component);
        fireEvent(container.getByTestId('flatlist_testid'), 'onEndReached');
    })
    it('flatlist renderItem check', () => {
        const container = render(component);
        const element = container.getByTestId('flatlist_testid');
        element.props.renderItem(fakeResponse);
    })
    it('refresh controller onRefresh check', async () => {
        const container = render(component);
        const element = container.getByTestId('flatlist_testid');
        const { refreshControl } = element.props;
        await act(async () => {
            refreshControl.props.onRefresh();
        });
    })
    it('check RESTAPI with response', async () => {
        const mockFetch = Promise.resolve({ json: () => Promise.resolve(fakeResponse) });
        const mockedFetch = jest.spyOn(API, 'RestAPI').mockImplementationOnce(() => mockFetch as any)
        render(component);
        expect(mockedFetch).toHaveBeenCalledTimes(1);
    })
    it('check RESTAPI without response', async () => {
        const mockFetch = Promise.resolve(undefined);
        const mockedFetch = jest.spyOn(API, 'RestAPI').mockImplementationOnce(() => mockFetch as any)
        render(component);
        expect(mockedFetch).toHaveBeenCalledTimes(3);
    })
})
