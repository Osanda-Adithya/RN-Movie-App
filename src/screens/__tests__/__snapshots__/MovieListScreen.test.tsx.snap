// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render successfully render snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        colors={
          Array [
            "red",
            "green",
            "blue",
          ]
        }
        onRefresh={[Function]}
        refreshing={true}
        title="Refreshing"
        titleColor="black"
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="flatlist_testid"
    viewabilityConfigCallbackPairs={Array []}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;
